server/pulp/server/db/model/__init__.py

```diff
@@ -536,6 +490,86 @@ def __hash__(self):
         return hash(self.unit_type_id + self.unit_key_str)
 
 
+class FileContentUnit(ContentUnit):
+    """
+    A content unit representing content that is of type *file* or *directory*.
+
+    :ivar _source_location: The absolute path to file or directory
+        to be copied to the platform storage location when the unit
+        is saved. See: set_content().
+    :type _source_location: str
+    """
+
+    def __init__(self, *args, **kwargs):
+        super(ContentUnit, self).__init__(*args, **kwargs)
+        self._source_location = None
+
+    @classmethod
+    def pre_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered before a unit is saved, this is used to
+        support the legacy behavior of generating the unit id and setting
+        the last_updated timestamp
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: FileContentUnit
+        """
+        super(FileContentUnit, cls).pre_save_signal(sender, document, **kwargs)
+        if not document._source_location:
+            # no content
+            return
+        with FileStorage() as storage:
+            storage.put(document, document._source_location)
+
+    def set_content(self, source_location):
+        """
+        Store the source of the content for the unit and the relative path
+        where it should be stored within the plugin content directory.
+
+        :param source_location: The absolute path to the content in the plugin working directory.
+        :type source_location: str
+
+        :raises PulpCodedException: PLP0036 if the source_location doesn't exist.
+        """
+        if not os.path.exists(source_location):
+            raise exceptions.PulpCodedException(error_code=error_codes.PLP0036,
+                                                source_location=source_location)
+        self._source_location = source_location
+
+
+class SharedContentUnit(ContentUnit):
+    """
+    A content unit representing content that is stored in a
+    shared storage facility.
+    """
+
+    @property
+    def storage_id(self):
```

Comment:
Good question.  Perhaps not requirements or limitations here but the SharedStorage should probably hash it so that it is of deterministic makeup and length.